File: bot/social_media/transcription/correction.py
"""Fuzzy-Korrektur eines Whisper-Transkripts gegen das Deadlock-Vokabular.

Vorgehen:
1. Aus dem Vokabular wird ein Index aus (term, alias) -> canonical aufgebaut.
2. Tokens des Transkripts werden zuerst exakt (case-insensitive) gegen den Index
   geprueft. Findet sich ein Treffer, wird der Token durch das `canonical`
   ersetzt.
3. Schlaegt der exakte Match fehl, wird per Levenshtein-Distanz auf das gesamte
   Vokabular gesucht. Treffer mit Distanz <= adaptivem Schwellwert (laenger =
   tolerant) werden ersetzt.
4. Multi-Word-Aliase (z.B. "soul orb") werden in einer zweiten Pass-Phase als
   N-Grams (bigrams + trigrams) gegen den Index geprueft, bevor Token-by-Token
   korrigiert wird.

Rueckgabe enthaelt das korrigierte Transkript und alle erkannten kanonischen
Begriffe (deduped, sortiert nach Vorkommen-Reihenfolge), die spaeter dem LLM
als Domain-Hinweis uebergeben werden.
"""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass, field
from typing import Iterable, Sequence

from .vocab import VocabEntry, load_all_vocab

log = logging.getLogger("TwitchStreams.SocialMedia.Correction")

_TOKEN_RE = re.compile(r"[A-Za-zÄÖÜäöüß']+", re.UNICODE)


@dataclass(frozen=True)
class CorrectionResult:
    corrected: str
    detected_terms: tuple[str, ...] = field(default_factory=tuple)
    replacements: tuple[tuple[str, str], ...] = field(default_factory=tuple)


# ---------- Levenshtein ----------

def _levenshtein(a: str, b: str) -> int:
    """Iterative Levenshtein with O(min(|a|,|b|)) memory."""
    if a == b:
        return 0
    if len(a) < len(b):
        a, b = b, a
    if not b:
        return len(a)
    previous = list(range(len(b) + 1))
    for i, ca in enumerate(a, start=1):
        current = [i] + [0] * len(b)
        for j, cb in enumerate(b, start=1):
            insert_cost = current[j - 1] + 1
            delete_cost = previous[j] + 1
            replace_cost = previous[j - 1] + (0 if ca == cb else 1)
            current[j] = min(insert_cost, delete_cost, replace_cost)
        previous = current
    return previous[-1]


def _adaptive_threshold(token_length: int) -> int:
    """Toleranzschwelle abhaengig von Token-Laenge (kuerzere Tokens strenger)."""
    if token_length <= 3:
        return 0
    if token_length <= 5:
        return 1
    if token_length <= 8:
        return 2
    return 2


# ---------- Index-Aufbau ----------

@dataclass(frozen=True)
class _VocabIndex:
    exact_lookup: dict[str, str]
    multi_word_lookup: dict[str, str]
    single_word_terms: tuple[tuple[str, str, int], ...]


def _build_index(entries: Sequence[VocabEntry]) -> _VocabIndex:
    exact: dict[str, str] = {}
    multi: dict[str, str] = {}
    singles: list[tuple[str, str, int]] = []
    for entry in entries:
        canonical = entry.canonical.strip()
        if not canonical:
            continue
        candidates = [entry.term, canonical, *entry.aliases]
        for cand in candidates:
            cand_text = str(cand or "").strip()
            if not cand_text:
                continue
            cand_lower = cand_text.lower()
            tokens = _TOKEN_RE.findall(cand_lower)
            if not tokens:
                continue
            if len(tokens) == 1:
                # exact lookup overrides only if heavier weight or new
                existing_weight = -1
                if cand_lower in exact:
                    # Hold first wins, no weight juggling needed
                    continue
                exact[cand_lower] = canonical
                singles.append((cand_lower, canonical, int(entry.weight or 1)))
                _ = existing_weight
            else:
                multi[" ".join(tokens)] = canonical

    # dedupe singles, keep highest weight
    dedup_singles: dict[str, tuple[str, int]] = {}
    for token, canon, weight in singles:
        prev = dedup_singles.get(token)
        if prev is None or weight > prev[1]:
            dedup_singles[token] = (canon, weight)
    singles_tuple = tuple(
        (token, canon, weight) for token, (canon, weight) in dedup_singles.items()
    )
    return _VocabIndex(
        exact_lookup=exact,
        multi_word_lookup=multi,
        single_word_terms=singles_tuple,
    )


# ---------- Korrektur ----------

def _replace_multi_word(text: str, index: _VocabIndex) -> tuple[str, list[str]]:
    """Ersetze Multi-Word-Aliasse (Bigrams + Trigrams) im Volltext."""
    if not index.multi_word_lookup:
        return text, []

    detected: list[str] = []

    # Sort patterns by length DESC to prefer longer matches first
    patterns = sorted(index.multi_word_lookup.items(), key=lambda x: -len(x[0]))
    for phrase, canonical in patterns:
        # Word-boundary regex with whitespace tolerance
        pattern = re.compile(
            r"\b" + r"\s+".join(re.escape(p) for p in phrase.split()) + r"\b",
            re.IGNORECASE,
        )
        if not pattern.search(text):
            continue
        text, count = pattern.subn(canonical, text)
        if count > 0:
            detected.extend([canonical] * count)
    return text, detected


def _correct_token(token: str, index: _VocabIndex) -> tuple[str, str | None]:
    """Korrigiere einen Token, gib (replacement, canonical_or_None) zurueck."""
    if not token or not token.isalpha():
        return token, None
    lower = token.lower()
    canonical = index.exact_lookup.get(lower)
    if canonical:
        return canonical, canonical

    threshold = _adaptive_threshold(len(lower))
    if threshold <= 0:
        return token, None

    best_distance = threshold + 1
    best_canonical: str | None = None
    best_weight = -1
    for term, canon, weight in index.single_word_terms:
        if abs(len(term) - len(lower)) > threshold:
            continue
        dist = _levenshtein(lower, term)
        if dist < best_distance or (dist == best_distance and weight > best_weight):
            best_distance = dist
            best_canonical = canon
            best_weight = weight
    if best_canonical and best_distance <= threshold:
        return best_canonical, best_canonical
    return token, None


def correct_transcript(
    transcript: str,
    *,
    vocab: Iterable[VocabEntry] | None = None,
) -> CorrectionResult:
    """Korrigiere das Transkript und sammle erkannte Domain-Begriffe.

    Args:
        transcript: Whisper-Rohtext.
        vocab: Optionaler Vorab-geladener VocabEntry-Iterable. Wenn None, wird
            das gesamte Vokabular aus der DB geladen (Caching dem Aufrufer
            ueberlassen).
    """
    if not transcript or not transcript.strip():
        return CorrectionResult(corrected=transcript or "")

    entries = list(vocab) if vocab is not None else load_all_vocab()
    if not entries:
        return CorrectionResult(corrected=transcript)

    index = _build_index(entries)

    text = transcript
    text, multi_detected = _replace_multi_word(text, index)

    detected: list[str] = list(multi_detected)
    replacements: list[tuple[str, str]] = []

    def _process_token(match: re.Match[str]) -> str:
        original = match.group(0)
        replacement, canonical = _correct_token(original, index)
        if canonical and replacement.lower() != original.lower():
            replacements.append((original, replacement))
        if canonical:
            detected.append(canonical)
        return replacement

    corrected = _TOKEN_RE.sub(_process_token, text)

    # dedupe detected (preserve order)
    seen: set[str] = set()
    deduped: list[str] = []
    for term in detected:
        if term in seen:
            continue
        seen.add(term)
        deduped.append(term)

    return CorrectionResult(
        corrected=corrected,
        detected_terms=tuple(deduped),
        replacements=tuple(replacements),
    )

File: rust/crates/tb-social-media/src/correction.rs
//! Fuzzy-Korrektur eines Whisper-Transkripts gegen das Deadlock-Vokabular
//! (Port von `bot/social_media/transcription/correction.py`).
//!
//! 1. Index aus (term, alias) → canonical bauen.
//! 2. Multi-Word-Aliase (Bigrams/Trigrams) als Phrasen im Volltext ersetzen.
//! 3. Token-für-Token: exakter Lookup, sonst Levenshtein ≤ adaptiver Schwelle.
//! Liefert korrigierten Text + erkannte canonical-Begriffe (dedupe, in
//! Vorkommens-Reihenfolge) für den späteren LLM-Domain-Hinweis. Reine
//! Textverarbeitung (Vokabular lädt der Aufrufer via [`crate::vocab`]).

use std::collections::HashMap;
use std::sync::OnceLock;

use regex::{Captures, NoExpand, Regex};

use crate::vocab::VocabEntry;

/// Ergebnis der Korrektur.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CorrectionResult {
    pub corrected: String,
    pub detected_terms: Vec<String>,
    pub replacements: Vec<(String, String)>,
}

/// Token-Regex (Python `_TOKEN_RE`): Buchstaben (inkl. Umlaute/ß) + Apostroph.
fn token_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"[A-Za-zÄÖÜäöüß']+").unwrap())
}

/// Iterative Levenshtein-Distanz, O(min(|a|,|b|)) Speicher.
fn levenshtein(a: &str, b: &str) -> usize {
    if a == b {
        return 0;
    }
    let (a, b): (Vec<char>, Vec<char>) = if a.chars().count() < b.chars().count() {
        (b.chars().collect(), a.chars().collect())
    } else {
        (a.chars().collect(), b.chars().collect())
    };
    if b.is_empty() {
        return a.len();
    }
    let mut previous: Vec<usize> = (0..=b.len()).collect();
    for (i, ca) in a.iter().enumerate() {
        let mut current = vec![0usize; b.len() + 1];
        current[0] = i + 1;
        for (j, cb) in b.iter().enumerate() {
            let insert = current[j] + 1;
            let delete = previous[j + 1] + 1;
            let replace = previous[j] + usize::from(ca != cb);
            current[j + 1] = insert.min(delete).min(replace);
        }
        previous = current;
    }
    previous[b.len()]
}

/// Toleranzschwelle abhängig von Token-Länge (Python `_adaptive_threshold`).
fn adaptive_threshold(token_length: usize) -> usize {
    match token_length {
        0..=3 => 0,
        4..=5 => 1,
        _ => 2,
    }
}

/// Vokabular-Index für die Korrektur.
struct VocabIndex {
    exact: HashMap<String, String>,
    multi: HashMap<String, String>,
    singles: Vec<(String, String, i32)>,
}

/// Tokenisiert kleingeschrieben (für den Index-Aufbau).
fn tokens_lower(text: &str) -> Vec<String> {
    let lower = text.to_lowercase();
    token_re().find_iter(&lower).map(|m| m.as_str().to_string()).collect()
}

fn build_index(entries: &[VocabEntry]) -> VocabIndex {
    let mut exact: HashMap<String, String> = HashMap::new();
    let mut multi: HashMap<String, String> = HashMap::new();
    let mut singles: Vec<(String, String, i32)> = Vec::new();

    for entry in entries {
        let canonical = entry.canonical.trim();
        if canonical.is_empty() {
            continue;
        }
        let mut candidates: Vec<&str> = vec![entry.term.as_str(), canonical];
        candidates.extend(entry.aliases.iter().map(String::as_str));
        for cand in candidates {
            let cand_text = cand.trim();
            if cand_text.is_empty() {
                continue;
            }
            let tokens = tokens_lower(cand_text);
            if tokens.is_empty() {
                continue;
            }
            if tokens.len() == 1 {
                let key = tokens[0].clone();
                // exact: first-wins.
                exact.entry(key.clone()).or_insert_with(|| canonical.to_string());
                singles.push((key, canonical.to_string(), entry.weight.max(1)));
            } else {
                multi.insert(tokens.join(" "), canonical.to_string());
            }
        }
    }

    // singles dedup: höchstes Gewicht behalten.
    let mut dedup: HashMap<String, (String, i32)> = HashMap::new();
    for (token, canon, weight) in singles {
        match dedup.get(&token) {
            Some((_, w)) if *w >= weight => {}
            _ => {
                dedup.insert(token, (canon, weight));
            }
        }
    }
    let singles = dedup.into_iter().map(|(t, (c, w))| (t, c, w)).collect();

    VocabIndex { exact, multi, singles }
}

/// Ersetzt Multi-Word-Phrasen (längste zuerst, Word-Boundary, Whitespace-tolerant).
fn replace_multi_word(text: &str, index: &VocabIndex) -> (String, Vec<String>) {
    if index.multi.is_empty() {
        return (text.to_string(), Vec::new());
    }
    let mut patterns: Vec<(&String, &String)> = index.multi.iter().collect();
    patterns.sort_by_key(|(phrase, _)| std::cmp::Reverse(phrase.len()));

    let mut text = text.to_string();
    let mut detected = Vec::new();
    for (phrase, canonical) in patterns {
        let escaped: Vec<String> = phrase.split_whitespace().map(regex::escape).collect();
        let pat = format!(r"(?i)\b{}\b", escaped.join(r"\s+"));
        let Ok(re) = Regex::new(&pat) else { continue };
        let count = re.find_iter(&text).count();
        if count == 0 {
            continue;
        }
        text = re.replace_all(&text, NoExpand(canonical)).into_owned();
        for _ in 0..count {
            detected.push(canonical.clone());
        }
    }
    (text, detected)
}

/// Korrigiert einen einzelnen Token → (Ersetzung, canonical | None).
fn correct_token(token: &str, index: &VocabIndex) -> (String, Option<String>) {
    if token.is_empty() || !token.chars().all(char::is_alphabetic) {
        return (token.to_string(), None);
    }
    let lower = token.to_lowercase();
    if let Some(canonical) = index.exact.get(&lower) {
        return (canonical.clone(), Some(canonical.clone()));
    }
    let threshold = adaptive_threshold(lower.chars().count());
    if threshold == 0 {
        return (token.to_string(), None);
    }
    let lower_len = lower.chars().count();
    let mut best_distance = threshold + 1;
    let mut best_canonical: Option<&str> = None;
    let mut best_weight = -1;
    for (term, canon, weight) in &index.singles {
        let term_len = term.chars().count();
        if term_len.abs_diff(lower_len) > threshold {
            continue;
        }
        let dist = levenshtein(&lower, term);
        if dist < best_distance || (dist == best_distance && *weight > best_weight) {
            best_distance = dist;
            best_canonical = Some(canon);
            best_weight = *weight;
        }
    }
    match best_canonical {
        Some(c) if best_distance <= threshold => (c.to_string(), Some(c.to_string())),
        _ => (token.to_string(), None),
    }
}

/// Korrigiert das Transkript + sammelt erkannte Domain-Begriffe (Python
/// `correct_transcript`). Leere/vokabularlose Eingabe → unverändert.
pub fn correct_transcript(transcript: &str, vocab: &[VocabEntry]) -> CorrectionResult {
    if transcript.trim().is_empty() || vocab.is_empty() {
        return CorrectionResult {
            corrected: transcript.to_string(),
            detected_terms: Vec::new(),
            replacements: Vec::new(),
        };
    }
    let index = build_index(vocab);
    let (text, multi_detected) = replace_multi_word(transcript, &index);

    let mut detected = multi_detected;
    let mut replacements: Vec<(String, String)> = Vec::new();
    let corrected = token_re()
        .replace_all(&text, |caps: &Captures| {
            let original = &caps[0];
            let (replacement, canonical) = correct_token(original, &index);
            if let Some(c) = canonical {
                if replacement.to_lowercase() != original.to_lowercase() {
                    replacements.push((original.to_string(), replacement.clone()));
                }
                detected.push(c);
            }
            replacement
        })
        .into_owned();

    // detected dedupe, Reihenfolge erhalten.
    let mut seen = std::collections::HashSet::new();
    let detected_terms: Vec<String> = detected.into_iter().filter(|t| seen.insert(t.clone())).collect();

    CorrectionResult { corrected, detected_terms, replacements }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn entry(term: &str, canonical: &str, aliases: &[&str], weight: i32) -> VocabEntry {
        VocabEntry {
            term: term.to_string(),
            canonical: canonical.to_string(),
            category: "hero".to_string(),
            source: "manual".to_string(),
            aliases: aliases.iter().map(|s| s.to_string()).collect(),
            weight,
            updated_at: None,
        }
    }

    #[test]
    fn levenshtein_und_threshold() {
        assert_eq!(levenshtein("haze", "haze"), 0);
        assert_eq!(levenshtein("haze", "hase"), 1);
        assert_eq!(levenshtein("", "abc"), 3);
        assert_eq!(adaptive_threshold(3), 0);
        assert_eq!(adaptive_threshold(5), 1);
        assert_eq!(adaptive_threshold(9), 2);
    }

    #[test]
    fn exakte_und_fuzzy_korrektur() {
        let vocab = vec![entry("haze", "Haze", &[], 5), entry("bebop", "Bebop", &[], 5)];
        let r = correct_transcript("ich spiele haze und bebob heute", &vocab);
        // "haze" exakt → Haze; "bebob" fuzzy (dist 1, len 5 → thresh 1) → Bebop.
        assert!(r.corrected.contains("Haze"));
        assert!(r.corrected.contains("Bebop"));
        assert!(r.detected_terms.contains(&"Haze".to_string()));
        assert!(r.detected_terms.contains(&"Bebop".to_string()));
        // Replacements erfasst (Original→Ersetzung), nur bei echter Änderung.
        assert!(r.replacements.iter().any(|(o, n)| o == "bebob" && n == "Bebop"));
    }

    #[test]
    fn multi_word_alias() {
        let mut e = entry("soul_orb", "Soul Orb", &["soul orb"], 5);
        e.term = "soul_orb".to_string();
        let vocab = vec![e];
        let r = correct_transcript("hol dir die soul   orb schnell", &vocab);
        assert!(r.corrected.contains("Soul Orb"));
        assert!(r.detected_terms.contains(&"Soul Orb".to_string()));
    }

    #[test]
    fn kurzer_token_nicht_fuzzy_korrigiert() {
        // "ich" (len 3 → thresh 0) wird nicht gegen "ice" gematcht.
        let vocab = vec![entry("ice", "Ice", &[], 5)];
        let r = correct_transcript("ich bin hier", &vocab);
        assert_eq!(r.corrected, "ich bin hier");
        assert!(r.detected_terms.is_empty());
    }

    #[test]
    fn leeres_oder_kein_vokabular() {
        assert_eq!(correct_transcript("   ", &[]).corrected, "   ");
        let r = correct_transcript("text", &[]);
        assert_eq!(r.corrected, "text");
        assert!(r.detected_terms.is_empty());
    }
}
